File: clients/haskell/src/Model/VehicleOrder.hs
module Model.VehicleOrder where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)


-- | Order for controlling a single vehicle
data VehicleOrder = VehicleOrder {
    -- | Acceleration (-1 - fully backwards, +1 - fully forward)
    accelerate :: Double,
    -- | Hand brakes
    brakes :: Bool,
    -- | Rotate the steering wheel (-1 - full clockwise, +1 - full counterclockwise)
    rotate :: Double }
    deriving Show

instance Trans VehicleOrder where
    read = do
        accelerate <- Trans.read
        brakes <- Trans.read
        rotate <- Trans.read
        return VehicleOrder {
            accelerate,
            brakes,
            rotate }
    
    write VehicleOrder {
        accelerate,
        brakes,
        rotate } = do
            Trans.write accelerate
            Trans.write brakes
            Trans.write rotate

File: clients/rust/src/model/vehicle_order.rs
use super::*;

/// Order for controlling a single vehicle
#[derive(Clone, Debug)]
pub struct VehicleOrder {
    /// Acceleration (-1 - fully backwards, +1 - fully forward)
    pub accelerate: f64,
    /// Hand brakes
    pub brakes: bool,
    /// Rotate the steering wheel (-1 - full clockwise, +1 - full counterclockwise)
    pub rotate: f64,
}

impl trans::Trans for VehicleOrder {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.accelerate.write_to(writer)?;
        self.brakes.write_to(writer)?;
        self.rotate.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let accelerate: f64 = trans::Trans::read_from(reader)?;
        let brakes: bool = trans::Trans::read_from(reader)?;
        let rotate: f64 = trans::Trans::read_from(reader)?;
        Ok(Self {
            accelerate,
            brakes,
            rotate,
        })
    }
}